File: crates/decant-core/src/stats.rs
use crate::Result;
use rusqlite::Connection;
use serde::Serialize;

#[derive(Debug, Serialize, PartialEq)]
pub struct Totals {
    pub sessions: i64,
    pub messages: i64,
    pub tool_calls: i64,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub cache_read_tokens: i64,
    pub cache_creation_tokens: i64,
    pub estimated_cost_usd: f64,
}

/// Whole-archive rollup.
pub fn totals(conn: &Connection) -> Result<Totals> {
    let t = conn.query_row(
        "SELECT
           (SELECT COUNT(*) FROM session),
           (SELECT COUNT(*) FROM message),
           (SELECT COUNT(*) FROM tool_call),
           (SELECT COALESCE(SUM(total_input_tokens),0) FROM session),
           (SELECT COALESCE(SUM(total_output_tokens),0) FROM session),
           (SELECT COALESCE(SUM(total_cache_read_tokens),0) FROM session),
           (SELECT COALESCE(SUM(total_cache_creation_tokens),0) FROM session),
           (SELECT COALESCE(SUM(estimated_cost_usd),0.0) FROM session)",
        [],
        |r| {
            Ok(Totals {
                sessions: r.get(0)?,
                messages: r.get(1)?,
                tool_calls: r.get(2)?,
                input_tokens: r.get(3)?,
                output_tokens: r.get(4)?,
                cache_read_tokens: r.get(5)?,
                cache_creation_tokens: r.get(6)?,
                estimated_cost_usd: r.get(7)?,
            })
        },
    )?;
    Ok(t)
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Dimension {
    Tool,
    Model,
    Project,
    Day,
}

impl Dimension {
    /// Parse the `--by` value. Returns None for unknown (caller reports a clear error).
    pub fn parse(s: &str) -> Option<Dimension> {
        match s {
            "tool" => Some(Dimension::Tool),
            "model" => Some(Dimension::Model),
            "project" => Some(Dimension::Project),
            "day" => Some(Dimension::Day),
            _ => None,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct DimRow {
    pub key: String,
    pub sessions: i64,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub estimated_cost_usd: f64,
}

/// Per-dimension rollup, ordered by session count desc. The grouping expression is
/// chosen from a fixed match (never user text) so the format! is injection-safe.
pub fn by_dimension(conn: &Connection, dim: Dimension) -> Result<Vec<DimRow>> {
    let (group_expr, join) = match dim {
        Dimension::Tool => ("s.tool", ""),
        Dimension::Model => ("COALESCE(s.model, '(unknown)')", ""),
        Dimension::Project => (
            "COALESCE(p.path, '(none)')",
            "LEFT JOIN project p ON p.id = s.project_id",
        ),
        Dimension::Day => ("substr(s.started_at, 1, 10)", ""),
    };
    let sql = format!(
        "SELECT {group_expr} AS k,
                COUNT(*) AS sessions,
                COALESCE(SUM(s.total_input_tokens),0),
                COALESCE(SUM(s.total_output_tokens),0),
                COALESCE(SUM(s.estimated_cost_usd),0.0)
         FROM session s {join}
         GROUP BY k
         ORDER BY sessions DESC"
    );
    let mut stmt = conn.prepare(&sql)?;
    let rows = stmt
        .query_map([], |r| {
            Ok(DimRow {
                key: r.get::<_, Option<String>>(0)?.unwrap_or_default(),
                sessions: r.get(1)?,
                input_tokens: r.get(2)?,
                output_tokens: r.get(3)?,
                estimated_cost_usd: r.get(4)?,
            })
        })?
        .collect::<std::result::Result<Vec<_>, _>>()?;
    Ok(rows)
}

#[derive(Debug, Serialize)]
pub struct ToolStatRow {
    pub tool_name: String,
    pub tool_kind: String,
    pub mcp_server: Option<String>,
    pub calls: i64,
    pub errors: i64,
}

/// Per-tool usage, most-called first. `errors_only` keeps only tools with >0 errors.
pub fn tool_usage(conn: &Connection, errors_only: bool, limit: i64) -> Result<Vec<ToolStatRow>> {
    let limit = if limit > 0 { limit } else { 50 };
    let having = if errors_only { "HAVING errors > 0" } else { "" };
    let sql = format!(
        "SELECT tool_name, tool_kind, mcp_server,
                COUNT(*) AS calls,
                COALESCE(SUM(CASE WHEN is_error = 1 THEN 1 ELSE 0 END), 0) AS errors
         FROM tool_call
         GROUP BY tool_name, tool_kind, mcp_server
         {having}
         ORDER BY calls DESC
         LIMIT {limit}"
    );
    let mut stmt = conn.prepare(&sql)?;
    let rows = stmt
        .query_map([], |r| {
            Ok(ToolStatRow {
                tool_name: r.get::<_, Option<String>>(0)?.unwrap_or_default(),
                tool_kind: r.get::<_, Option<String>>(1)?.unwrap_or_default(),
                mcp_server: r.get(2)?,
                calls: r.get(3)?,
                errors: r.get(4)?,
            })
        })?
        .collect::<std::result::Result<Vec<_>, _>>()?;
    Ok(rows)
}

#[derive(Debug, Serialize)]
pub struct McpStatRow {
    pub mcp_server: String,
    pub tools: i64,
    pub calls: i64,
    pub errors: i64,
}

/// Per-MCP-server usage, most-called first.
pub fn mcp_usage(conn: &Connection, limit: i64) -> Result<Vec<McpStatRow>> {
    let limit = if limit > 0 { limit } else { 50 };
    let mut stmt = conn.prepare(
        "SELECT mcp_server,
                COUNT(DISTINCT tool_name) AS tools,
                COUNT(*) AS calls,
                COALESCE(SUM(CASE WHEN is_error = 1 THEN 1 ELSE 0 END), 0) AS errors
         FROM tool_call
         WHERE tool_kind = 'mcp' AND mcp_server IS NOT NULL
         GROUP BY mcp_server
         ORDER BY calls DESC
         LIMIT ?1",
    )?;
    let rows = stmt
        .query_map([limit], |r| {
            Ok(McpStatRow {
                mcp_server: r.get::<_, Option<String>>(0)?.unwrap_or_default(),
                tools: r.get(1)?,
                calls: r.get(2)?,
                errors: r.get(3)?,
            })
        })?
        .collect::<std::result::Result<Vec<_>, _>>()?;
    Ok(rows)
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FileGroup {
    /// Hotspot files, keyed by project-relative path.
    Path,
    /// Language lens, keyed by file extension.
    Ext,
}

impl FileGroup {
    pub fn parse(s: &str) -> Option<FileGroup> {
        match s {
            "path" => Some(FileGroup::Path),
            "ext" => Some(FileGroup::Ext),
            _ => None,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct FileStatRow {
    /// rel_path (path fallback) for `Path`, extension for `Ext`.
    pub key: String,
    /// Owning project path (`Path` grouping only).
    pub project: Option<String>,
    pub reads: i64,
    pub edits: i64,
    pub writes: i64,
    pub deletes: i64,
    pub sessions: i64,
    pub last_touched_at: Option<String>,
}

/// File hotspots: which files (or extensions) the agents touch most, split by
/// operation. Ordered by total operations desc. The grouping expression comes
/// from a fixed match (never user text), so the format! is injection-safe.
pub fn file_hotspots(
    conn: &Connection,
    group: FileGroup,
    op: Option<crate::enrich::Operation>,
    limit: i64,
) -> Result<Vec<FileStatRow>> {
    let limit = if limit > 0 { limit } else { 50 };
    let (key_expr, proj_expr, join) = match group {
        FileGroup::Path => (
            "COALESCE(f.rel_path, f.path)",
            "p.path",
            "JOIN session s ON s.id = f.session_id LEFT JOIN project p ON p.id = s.project_id",
        ),
        FileGroup::Ext => (
            "COALESCE(f.ext, '(none)')",
            "NULL",
            "JOIN session s ON s.id = f.session_id",
        ),
    };
    let op_filter = match op {
        Some(_) => "WHERE f.operation = ?1",
        None => "",
    };
    let sql = format!(
        "SELECT {key_expr} AS k, {proj_expr} AS proj,
                SUM(f.operation = 'read') AS reads,
                SUM(f.operation = 'edit') AS edits,
                SUM(f.operation = 'write') AS writes,
                SUM(f.operation = 'delete') AS deletes,
                COUNT(DISTINCT f.session_id) AS sessions,
                MAX(f.timestamp) AS last_touched
         FROM file_ref f {join}
         {op_filter}
         GROUP BY k, proj
         ORDER BY (reads + edits + writes + deletes) DESC, k ASC
         LIMIT {limit}"
    );
    let mut stmt = conn.prepare(&sql)?;
    let map_row = |r: &rusqlite::Row<'_>| {
        Ok(FileStatRow {
            key: r.get::<_, Option<String>>(0)?.unwrap_or_default(),
            project: r.get(1)?,
            reads: r.get(2)?,
            edits: r.get(3)?,
            writes: r.get(4)?,
            deletes: r.get(5)?,
            sessions: r.get(6)?,
            last_touched_at: r.get(7)?,
        })
    };
    let rows = match op {
        Some(o) => stmt.query_map([o.as_str()], map_row)?,
        None => stmt.query_map([], map_row)?,
    }
    .collect::<std::result::Result<Vec<_>, _>>()?;
    Ok(rows)
}

/// One session's facet counters + classification, for CLI `show`.
#[derive(Debug, Serialize)]
pub struct SessionFacetRow {
    pub turn_count: i64,
    pub error_count: i64,
    pub interruption_count: i64,
    pub compaction_count: i64,
    pub sidechain_message_count: i64,
    pub agent_spawn_count: i64,
    pub skill_count: i64,
    pub command_count: i64,
    pub thinking_block_count: i64,
    pub thinking_chars: i64,
    pub active_seconds: i64,
    pub outcome: Option<String>,
    pub work_type: Option<String>,
}

/// Fetch the deterministic facets for one session (None if the id is unknown).
pub fn session_facets(conn: &Connection, session_id: i64) -> Result<Option<SessionFacetRow>> {
    let row = conn
        .query_row(
            "SELECT turn_count, error_count, interruption_count, compaction_count,
                    sidechain_message_count, agent_spawn_count, skill_count, command_count,
                    thinking_block_count, thinking_chars, active_seconds, outcome, work_type
             FROM session WHERE id = ?1",
            [session_id],
            |r| {
                Ok(SessionFacetRow {
                    turn_count: r.get(0)?,
                    error_count: r.get(1)?,
                    interruption_count: r.get(2)?,
                    compaction_count: r.get(3)?,
                    sidechain_message_count: r.get(4)?,
                    agent_spawn_count: r.get(5)?,
                    skill_count: r.get(6)?,
                    command_count: r.get(7)?,
                    thinking_block_count: r.get(8)?,
                    thinking_chars: r.get(9)?,
                    active_seconds: r.get(10)?,
                    outcome: r.get(11)?,
                    work_type: r.get(12)?,
                })
            },
        )
        .map(Some)
        .or_else(|e| match e {
            rusqlite::Error::QueryReturnedNoRows => Ok(None),
            other => Err(other),
        })?;
    Ok(row)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{db, ingest, schema, sources};

    fn seeded() -> Connection {
        let conn = db::open_in_memory().unwrap();
        schema::migrate(&conn).unwrap();
        let content = std::fs::read_to_string(concat!(
            env!("CARGO_MANIFEST_DIR"),
            "/../../fixtures/claude/sample.jsonl"
        ))
        .unwrap();
        let parsed = sources::claude::parse_session("sess-claude-1", &content);
        let tx = conn.unchecked_transaction().unwrap();
        ingest::upsert_session(&tx, &parsed, "/x.jsonl", 1, 2, "h").unwrap();
        tx.commit().unwrap();
        conn
    }

    fn seeded_enriched() -> Connection {
        let conn = db::open_in_memory().unwrap();
        schema::migrate(&conn).unwrap();
        for (fixture, id) in [
            ("/../../fixtures/claude/enriched.jsonl", "claude"),
            ("/../../fixtures/codex/enriched.jsonl", "codex"),
        ] {
            let content =
                std::fs::read_to_string(format!("{}{}", env!("CARGO_MANIFEST_DIR"), fixture))
                    .unwrap();
            let parsed = if id == "claude" {
                sources::claude::parse_session("sess-enr-claude", &content)
            } else {
                sources::codex::parse_session("sess-enr-codex", &content, &Default::default())
            };
            let tx = conn.unchecked_transaction().unwrap();
            ingest::upsert_session(&tx, &parsed, &format!("/x/{id}.jsonl"), 1, 2, id).unwrap();
            tx.commit().unwrap();
        }
        conn
    }

    #[test]
    fn file_hotspots_by_path_orders_by_total_ops() {
        let conn = seeded_enriched();
        let rows = file_hotspots(&conn, FileGroup::Path, None, 50).unwrap();
        // claude: src/main.rs (read+edit), README.md (write), nb.ipynb (edit)
        // codex: docs/new.md (write), src/lib.rs (edit), old.txt (delete)
        assert_eq!(rows[0].key, "src/main.rs");
        assert_eq!((rows[0].reads, rows[0].edits), (1, 1));
        assert_eq!(rows[0].sessions, 1);
        assert_eq!(rows[0].project.as_deref(), Some("/Users/dev/proj"));
        assert_eq!(rows.len(), 6);
        assert!(rows[0].last_touched_at.is_some());
    }

    #[test]
    fn file_hotspots_op_filter_keeps_only_that_operation() {
        let conn = seeded_enriched();
        let rows = file_hotspots(
            &conn,
            FileGroup::Path,
            Some(crate::enrich::Operation::Edit),
            50,
        )
        .unwrap();
        let keys: Vec<_> = rows.iter().map(|r| r.key.as_str()).collect();
        assert_eq!(keys, vec!["nb.ipynb", "src/lib.rs", "src/main.rs"]);
        assert!(rows.iter().all(|r| r.reads == 0 && r.writes == 0));
    }

    #[test]
    fn file_hotspots_by_ext_rolls_up_languages() {
        let conn = seeded_enriched();
        let rows = file_hotspots(&conn, FileGroup::Ext, None, 50).unwrap();
        let rs = rows.iter().find(|r| r.key == "rs").unwrap();
        // src/main.rs read+edit (claude) + src/lib.rs edit (codex)
        assert_eq!((rs.reads, rs.edits), (1, 2));
        assert_eq!(rs.sessions, 2, "rs ops span both sessions");
        assert!(rs.project.is_none());
    }

    #[test]
    fn file_group_parse() {
        assert_eq!(FileGroup::parse("path"), Some(FileGroup::Path));
        assert_eq!(FileGroup::parse("ext"), Some(FileGroup::Ext));
        assert_eq!(FileGroup::parse("bogus"), None);
    }

    #[test]
    fn totals_rollup() {
        let conn = seeded();
        let t = totals(&conn).unwrap();
        assert_eq!(t.sessions, 1);
        assert_eq!(t.messages, 4);
        assert_eq!(t.tool_calls, 1);
        assert_eq!(t.input_tokens, 2700);
    }

    #[test]
    fn by_tool_and_day() {
        let conn = seeded();
        let by_tool = by_dimension(&conn, Dimension::Tool).unwrap();
        assert_eq!(by_tool.len(), 1);
        assert_eq!(by_tool[0].key, "claude_code");
        assert_eq!(by_tool[0].sessions, 1);

        let by_day = by_dimension(&conn, Dimension::Day).unwrap();
        assert_eq!(by_day[0].key, "2026-05-01");
    }

    #[test]
    fn dimension_parse() {
        assert_eq!(Dimension::parse("model"), Some(Dimension::Model));
        assert_eq!(Dimension::parse("nope"), None);
    }

    #[test]
    fn tool_usage_counts_builtin_call() {
        let conn = seeded();
        let tools = tool_usage(&conn, false, 50).unwrap();
        // fixture has one tool_use: builtin "Read"
        let read = tools.iter().find(|t| t.tool_name == "Read").unwrap();
        assert_eq!(read.tool_kind, "builtin");
        assert_eq!(read.calls, 1);
        assert_eq!(read.errors, 0);

        // errors_only filters it out (no errors in the fixture)
        assert!(tool_usage(&conn, true, 50).unwrap().is_empty());
    }

    #[test]
    fn mcp_usage_empty_for_builtin_only_fixture() {
        let conn = seeded();
        assert!(mcp_usage(&conn, 50).unwrap().is_empty());
    }

    #[test]
    fn by_dimension_project_and_model_group_correctly() {
        let conn = seeded();
        let by_project = by_dimension(&conn, Dimension::Project).unwrap();
        assert_eq!(by_project.len(), 1);
        assert_eq!(by_project[0].key, "/Users/dev/proj");
        assert_eq!(by_project[0].sessions, 1);

        let by_model = by_dimension(&conn, Dimension::Model).unwrap();
        assert_eq!(by_model[0].key, "claude-opus-4-7");
    }

    #[test]
    fn by_dimension_project_uses_placeholder_when_unlinked() {
        // A session with no project_id rolls up under "(none)".
        let conn = db::open_in_memory().unwrap();
        schema::migrate(&conn).unwrap();
        conn.execute_batch(
            "INSERT INTO session(id, tool, source_session_id) VALUES (1, 'codex', 's1');",
        )
        .unwrap();
        let by_project = by_dimension(&conn, Dimension::Project).unwrap();
        assert_eq!(by_project[0].key, "(none)");
    }

    #[test]
    fn session_facets_returns_row_for_known_id_and_none_for_unknown() {
        let conn = seeded_enriched();
        // The first ingested (claude) session has facet counters populated.
        let id: i64 = conn
            .query_row(
                "SELECT id FROM session WHERE source_session_id = 'sess-enr-claude'",
                [],
                |r| r.get(0),
            )
            .unwrap();
        let facets = session_facets(&conn, id).unwrap().expect("known session");
        assert_eq!(facets.turn_count, 1);
        assert!(facets.active_seconds > 0);

        // Unknown id → Ok(None), not an error.
        assert!(session_facets(&conn, 999_999).unwrap().is_none());
    }

    #[test]
    fn stat_queries_propagate_db_errors() {
        // An un-migrated DB lacks every table, so `totals`, `mcp_usage`, and
        // `session_facets` all surface the SQLite error via their `?`.
        let bare = db::open_in_memory().unwrap();
        assert!(totals(&bare).is_err());
        assert!(mcp_usage(&bare, 10).is_err());
        // session_facets' error is a non-NoRows SQLite error (no `session`
        // table), exercising its `other => Err(other)` arm.
        assert!(session_facets(&bare, 1).is_err());
    }
}

File: src/stats.ts
import type { Database } from "bun:sqlite";
import type { Operation } from "./enrich.ts";

export interface Totals {
  sessions: number;
  messages: number;
  tool_calls: number;
  input_tokens: number;
  output_tokens: number;
  cache_read_tokens: number;
  cache_creation_tokens: number;
  reasoning_tokens: number;
  est_reasoning_tokens: number;
  estimated_cost_usd: number;
}

export function totals(db: Database): Totals {
  return db
    .query(
      `SELECT
         (SELECT COUNT(*) FROM session) AS sessions,
         (SELECT COUNT(*) FROM message) AS messages,
         (SELECT COUNT(*) FROM tool_call) AS tool_calls,
         (SELECT COALESCE(SUM(total_input_tokens), 0) FROM session) AS input_tokens,
         (SELECT COALESCE(SUM(total_output_tokens), 0) FROM session) AS output_tokens,
         (SELECT COALESCE(SUM(total_cache_read_tokens), 0) FROM session) AS cache_read_tokens,
         (SELECT COALESCE(SUM(total_cache_creation_tokens), 0) FROM session) AS cache_creation_tokens,
         (SELECT COALESCE(SUM(total_reasoning_tokens), 0) FROM session) AS reasoning_tokens,
         (SELECT COALESCE(SUM(est_reasoning_tokens), 0) FROM session) AS est_reasoning_tokens,
         (SELECT COALESCE(SUM(estimated_cost_usd), 0.0) FROM session) AS estimated_cost_usd`,
    )
    .get() as Totals;
}

export type Dimension = "tool" | "model" | "project" | "day";

export function parseDimension(value: string): Dimension | null {
  return value === "tool" || value === "model" || value === "project" || value === "day"
    ? value
    : null;
}

export interface DimRow {
  key: string;
  sessions: number;
  input_tokens: number;
  output_tokens: number;
  reasoning_tokens: number;
  est_reasoning_tokens: number;
  estimated_cost_usd: number;
}

interface DimRowDb extends Omit<DimRow, "key"> {
  key: string | null;
}

export function byDimension(db: Database, dimension: Dimension): DimRow[] {
  const { groupExpr, join } = dimensionSql(dimension);
  const rows = db
    .query(
      `SELECT ${groupExpr} AS key,
              COUNT(*) AS sessions,
              COALESCE(SUM(s.total_input_tokens), 0) AS input_tokens,
              COALESCE(SUM(s.total_output_tokens), 0) AS output_tokens,
              COALESCE(SUM(s.total_reasoning_tokens), 0) AS reasoning_tokens,
              COALESCE(SUM(s.est_reasoning_tokens), 0) AS est_reasoning_tokens,
              COALESCE(SUM(s.estimated_cost_usd), 0.0) AS estimated_cost_usd
       FROM session s ${join}
       GROUP BY key
       ORDER BY sessions DESC`,
    )
    .all() as DimRowDb[];
  return rows.map((row) => ({ ...row, key: row.key ?? "" }));
}

export interface ToolStatRow {
  tool_name: string;
  tool_kind: string;
  mcp_server: string | null;
  calls: number;
  errors: number;
}

interface ToolStatDb extends Omit<ToolStatRow, "tool_name" | "tool_kind"> {
  tool_name: string | null;
  tool_kind: string | null;
}

export function toolUsage(db: Database, errorsOnly: boolean, limitValue = 50): ToolStatRow[] {
  const limit = normalizeLimit(limitValue, 50);
  const having = errorsOnly ? "HAVING errors > 0" : "";
  const rows = db
    .query(
      `SELECT tool_name, tool_kind, mcp_server,
              COUNT(*) AS calls,
              COALESCE(SUM(CASE WHEN is_error = 1 THEN 1 ELSE 0 END), 0) AS errors
       FROM tool_call
       GROUP BY tool_name, tool_kind, mcp_server
       ${having}
       ORDER BY calls DESC
       LIMIT ${limit}`,
    )
    .all() as ToolStatDb[];
  return rows.map((row) => ({
    ...row,
    tool_name: row.tool_name ?? "",
    tool_kind: row.tool_kind ?? "",
  }));
}

export interface McpStatRow {
  mcp_server: string;
  tools: number;
  calls: number;
  errors: number;
}

interface McpStatDb extends Omit<McpStatRow, "mcp_server"> {
  mcp_server: string | null;
}

export function mcpUsage(db: Database, limitValue = 50): McpStatRow[] {
  const limit = normalizeLimit(limitValue, 50);
  const rows = db
    .query(
      `SELECT mcp_server,
              COUNT(DISTINCT tool_name) AS tools,
              COUNT(*) AS calls,
              COALESCE(SUM(CASE WHEN is_error = 1 THEN 1 ELSE 0 END), 0) AS errors
       FROM tool_call
       WHERE tool_kind = 'mcp' AND mcp_server IS NOT NULL
       GROUP BY mcp_server
       ORDER BY calls DESC
       LIMIT ?1`,
    )
    .all(limit) as McpStatDb[];
  return rows.map((row) => ({ ...row, mcp_server: row.mcp_server ?? "" }));
}

export type FileGroup = "path" | "ext";

export function parseFileGroup(value: string): FileGroup | null {
  return value === "path" || value === "ext" ? value : null;
}

export interface FileStatRow {
  key: string;
  project: string | null;
  reads: number;
  edits: number;
  writes: number;
  deletes: number;
  sessions: number;
  last_touched_at: string | null;
}

interface FileStatDb extends Omit<FileStatRow, "key"> {
  key: string | null;
}

export function fileHotspots(
  db: Database,
  group: FileGroup,
  op: Operation | null,
  limitValue = 50,
): FileStatRow[] {
  const limit = normalizeLimit(limitValue, 50);
  const { keyExpr, projectExpr, join } = fileGroupSql(group);
  const opFilter = op == null ? "" : "WHERE f.operation = ?1";
  const sql = `SELECT ${keyExpr} AS key, ${projectExpr} AS project,
                      SUM(f.operation = 'read') AS reads,
                      SUM(f.operation = 'edit') AS edits,
                      SUM(f.operation = 'write') AS writes,
                      SUM(f.operation = 'delete') AS deletes,
                      COUNT(DISTINCT f.session_id) AS sessions,
                      MAX(f.timestamp) AS last_touched_at
               FROM file_ref f ${join}
               ${opFilter}
               GROUP BY key, project
               ORDER BY (reads + edits + writes + deletes) DESC, key ASC
               LIMIT ${limit}`;
  const rows = (op == null ? db.query(sql).all() : db.query(sql).all(op)) as FileStatDb[];
  return rows.map((row) => ({ ...row, key: row.key ?? "" }));
}

export interface SessionFacetRow {
  turn_count: number;
  error_count: number;
  interruption_count: number;
  compaction_count: number;
  sidechain_message_count: number;
  agent_spawn_count: number;
  skill_count: number;
  command_count: number;
  thinking_block_count: number;
  thinking_chars: number;
  active_seconds: number;
  outcome: string | null;
  work_type: string | null;
}

export function sessionFacets(db: Database, sessionId: number): SessionFacetRow | null {
  return db
    .query(
      `SELECT turn_count, error_count, interruption_count, compaction_count,
              sidechain_message_count, agent_spawn_count, skill_count, command_count,
              thinking_block_count, thinking_chars, active_seconds, outcome, work_type
       FROM session WHERE id = ?1`,
    )
    .get(sessionId) as SessionFacetRow | null;
}

export interface Activity {
  by_hour: number[];
  by_weekday: number[];
  timezone: string;
  peak_hour: number | null;
  peak_weekday: number | null;
}

export function activity(db: Database): Activity {
  const byHour = bucket(db, "strftime('%H', s.started_at, 'localtime')", 24);
  const byWeekday = bucket(db, "strftime('%w', s.started_at, 'localtime')", 7);
  return {
    by_hour: byHour,
    by_weekday: byWeekday,
    timezone: localTimezone(),
    peak_hour: peak(byHour),
    peak_weekday: peak(byWeekday),
  };
}

export interface ModelSparklines {
  models: Record<string, number[]>;
  days: string[];
}

export function modelSparklines(db: Database): ModelSparklines {
  const rows = db
    .query(
      `SELECT COALESCE(s.model, '(unknown)') AS model,
              substr(s.started_at, 1, 10) AS day,
              COUNT(*) AS count
       FROM session s
       WHERE s.started_at IS NOT NULL
       GROUP BY model, day
       ORDER BY model ASC, day ASC`,
    )
    .all() as { model: string; day: string; count: number }[];
  const days = [...new Set(rows.map((row) => row.day))].sort();
  const dayIndex = new Map(days.map((day, index) => [day, index]));
  const models: Record<string, number[]> = {};
  for (const row of rows) {
    if (models[row.model] == null) {
      models[row.model] = Array.from({ length: days.length }, () => 0);
    }
    const modelCounts = models[row.model];
    const index = dayIndex.get(row.day);
    if (index != null && modelCounts != null) {
      modelCounts[index] = row.count;
    }
  }
  return { models, days };
}

export interface DateBounds {
  min: string | null;
  max: string | null;
}

export function dateBounds(db: Database): DateBounds {
  return db
    .query(
      `SELECT MIN(substr(started_at, 1, 10)) AS min,
              MAX(substr(started_at, 1, 10)) AS max
       FROM session
       WHERE started_at IS NOT NULL`,
    )
    .get() as DateBounds;
}

export function todayTotals(db: Database): Totals {
  return db
    .query(
      `SELECT
         COUNT(*) AS sessions,
         (SELECT COUNT(*) FROM message m JOIN session s2 ON s2.id = m.session_id
          WHERE substr(s2.started_at, 1, 10) = date('now', 'localtime')) AS messages,
         (SELECT COUNT(*) FROM tool_call t JOIN session s3 ON s3.id = t.session_id
          WHERE substr(s3.started_at, 1, 10) = date('now', 'localtime')) AS tool_calls,
         COALESCE(SUM(total_input_tokens), 0) AS input_tokens,
         COALESCE(SUM(total_output_tokens), 0) AS output_tokens,
         COALESCE(SUM(total_cache_read_tokens), 0) AS cache_read_tokens,
         COALESCE(SUM(total_cache_creation_tokens), 0) AS cache_creation_tokens,
         COALESCE(SUM(total_reasoning_tokens), 0) AS reasoning_tokens,
         COALESCE(SUM(est_reasoning_tokens), 0) AS est_reasoning_tokens,
         COALESCE(SUM(estimated_cost_usd), 0.0) AS estimated_cost_usd
       FROM session
       WHERE substr(started_at, 1, 10) = date('now', 'localtime')`,
    )
    .get() as Totals;
}

function dimensionSql(dimension: Dimension): { groupExpr: string; join: string } {
  switch (dimension) {
    case "tool":
      return { groupExpr: "s.tool", join: "" };
    case "model":
      return { groupExpr: "COALESCE(s.model, '(unknown)')", join: "" };
    case "project":
      return {
        groupExpr: "COALESCE(p.path, '(none)')",
        join: "LEFT JOIN project p ON p.id = s.project_id",
      };
    case "day":
      return { groupExpr: "substr(s.started_at, 1, 10)", join: "" };
  }
}

function fileGroupSql(group: FileGroup): { keyExpr: string; projectExpr: string; join: string } {
  switch (group) {
    case "path":
      return {
        keyExpr: "COALESCE(f.rel_path, f.path)",
        projectExpr: "p.path",
        join: "JOIN session s ON s.id = f.session_id LEFT JOIN project p ON p.id = s.project_id",
      };
    case "ext":
      return {
        keyExpr: "COALESCE(f.ext, '(none)')",
        projectExpr: "NULL",
        join: "JOIN session s ON s.id = f.session_id",
      };
  }
}

function bucket(db: Database, expr: string, size: number): number[] {
  const out = Array.from({ length: size }, () => 0);
  const rows = db
    .query(
      `SELECT ${expr} AS key, COUNT(*) AS count
       FROM session s
       WHERE s.started_at IS NOT NULL
       GROUP BY key`,
    )
    .all() as { key: string | null; count: number }[];
  for (const row of rows) {
    const key = row.key == null || row.key === "" ? Number.NaN : Number.parseInt(row.key, 10);
    if (Number.isInteger(key) && key >= 0 && key < size) {
      out[key] = row.count;
    }
  }
  return out;
}

function peak(counts: number[]): number | null {
  let bestIndex: number | null = null;
  let bestCount = 0;
  for (const [index, count] of counts.entries()) {
    if (count > bestCount) {
      bestIndex = index;
      bestCount = count;
    }
  }
  return bestIndex;
}

function localTimezone(): string {
  const offset = -new Date().getTimezoneOffset();
  const sign = offset < 0 ? "-" : "+";
  const absolute = Math.abs(offset);
  const hours = Math.floor(absolute / 60)
    .toString()
    .padStart(2, "0");
  const minutes = (absolute % 60).toString().padStart(2, "0");
  return `UTC${sign}${hours}:${minutes}`;
}

function normalizeLimit(value: number | null | undefined, fallback: number): number {
  return value != null && value > 0 ? Math.trunc(value) : fallback;
}
